### ISSUE TITLE
Add method to delete attachment to ArtifactStore

### ISSUE THREAD

**Issue by chetanmeh** [apache/incubator-openwhisk#3356]
As discussed on [mailing list][1] a method should be added to delete attachment to allow deletions in ArtifactStore implementation which store binary data in a separate store. To enable that

1. Add a method to ArtifactStore
```
  protected[core] def deleteAttachment[T](doc: DocInfo, name: String)(
    implicit transid: TransactionId): Future[Boolean]
```
2. Have this method exposed in `DocumentFactory`
3. Have the method in (2) invoked by `WhiskAction`
4. In `CouchDBStore` add a no op implementation which returns a successful future as delete is handled implicitly in `del` call for the entity

[1]: https://lists.apache.org/thread.html/b5c46924e9c9c951cb68492b1ecc101aacbccce2ab1cd74fc3bbdafa@%3Cdev.openwhisk.apache.org%3E

**Comment by chetanmeh** [comment]
Currently when any entity is deleted we do not have entity instance. So this would make it tricky to implement `deleteAttachment(docInfo, name)`. Instead we can have a method like

```
protected[core] def deleteAttachments[T](doc: DocInfo)(
    implicit transid: TransactionId): Future[Boolean]
```
Here the `ArtifactStore` implementation can remove all attachments related to the given docId. This requires that
1. custom `AttachmentStore` (probably a new SPI for later) stores attachment with blob id set to "${doc.id.id}/$name"
2. Delete operation can be implemented by removing all blobs with id starting with document id

This may involve multiple calls depending on `AttachmentStore`  implementation. 

### S3

S3 does not provide a direct call to [delete by prefix][1]. Instead we need to 

1. first [list blobid][2] matching given prefix 
2. Delete then via [multiple delete support][3]

### Azure Blob Storage

It does not support performing multiple deletes in single call. Instead we would need to 

1. [List Blobs][4] matching given prefix
2. Delete then [one by one][5]

[1]: https://stackoverflow.com/questions/3178685/how-to-delete-lot-of-objects-named-with-common-prefix-from-s3-bucket
[2]: https://docs.aws.amazon.com/AmazonS3/latest/API/v2-RESTBucketGET.html
[3]: https://docs.aws.amazon.com/AmazonS3/latest/API/multiobjectdeleteapi.html
[4]: https://docs.microsoft.com/en-us/rest/api/storageservices/list-blobs
[5]: https://docs.microsoft.com/en-us/rest/api/storageservices/delete-blob

**Comment by chetanmeh** [comment]
Closed via #3366
